File: backend/server/models/lab_reports.py
"""
LabReport — structured lab test results attached to a batch.

The pass/fail outcome (approved / rejected) lives in `BatchState.test_result`;
this table holds the structured metrics and notes for audit.
"""
from datetime import datetime

from . import db


class LabReport(db.Model):
    __tablename__ = "lab_reports"

    report_id = db.Column(db.String(50), primary_key=True)

    batch_id = db.Column(
        db.String(50),
        db.ForeignKey("herbs.batch_id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    lab_id = db.Column(
        db.String(50),
        db.ForeignKey("users.user_id"),
        nullable=False,
    )

    test_type = db.Column(db.String(100), nullable=False)
    test_date = db.Column(db.Date, nullable=False)
    results_summary = db.Column(db.Text, nullable=False)

    # Outcome
    outcome = db.Column(
        db.Enum("approved", "rejected", name="lab_outcome"),
        nullable=False,
    )
    certification_level = db.Column(db.String(50), nullable=True)

    # Quality metrics (all optional — different test types use different ones)
    purity_percentage = db.Column(db.Numeric(5, 2), nullable=True)
    moisture_content = db.Column(db.Numeric(5, 2), nullable=True)
    ash_content = db.Column(db.Numeric(5, 2), nullable=True)
    heavy_metals_present = db.Column(db.Boolean, nullable=True)
    pesticides_detected = db.Column(db.Boolean, nullable=True)

    # Medicinal properties
    active_compounds = db.Column(db.Text, nullable=True)
    potency_rating = db.Column(db.String(20), nullable=True)

    report_url = db.Column(db.String(500), nullable=True)
    notes = db.Column(db.Text, nullable=True)
    recommendations = db.Column(db.Text, nullable=True)

    created_at = db.Column(db.DateTime, nullable=False, default=datetime.utcnow)
    updated_at = db.Column(
        db.DateTime,
        nullable=False,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    lab = db.relationship("User", foreign_keys=[lab_id])

    def __repr__(self):
        return f"<LabReport {self.report_id} batch={self.batch_id} outcome={self.outcome}>"

    def to_dict(self) -> dict:
        return {
            "report_id": self.report_id,
            "batch_id": self.batch_id,
            "lab_id": self.lab_id,
            "test_type": self.test_type,
            "test_date": self.test_date.isoformat() if self.test_date else None,
            "results_summary": self.results_summary,
            "outcome": self.outcome,
            "certification_level": self.certification_level,
            "purity_percentage": float(self.purity_percentage) if self.purity_percentage is not None else None,
            "moisture_content": float(self.moisture_content) if self.moisture_content is not None else None,
            "ash_content": float(self.ash_content) if self.ash_content is not None else None,
            "heavy_metals_present": self.heavy_metals_present,
            "pesticides_detected": self.pesticides_detected,
            "active_compounds": self.active_compounds,
            "potency_rating": self.potency_rating,
            "report_url": self.report_url,
            "notes": self.notes,
            "recommendations": self.recommendations,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

File: second_backend/src/routes/labReports.js
/**
 * Lab report routes — port of routes/lab_reports.py.
 *
 *   POST /api/v1/lab-reports                  lab files a report (must be holder)
 *   GET  /api/v1/lab-reports/batch/<batch_id> list reports for a batch
 *   GET  /api/v1/lab-reports/<report_id>      get a single report
 */
const express = require("express");
const { prisma } = require("../db/client");
const { asyncHandler } = require("../utils/asyncHandler");
const { requireAuth, requireRole } = require("../middleware/auth");
const { ok, error } = require("../utils/responses");
const { createLabReportSchema, zodDetails } = require("../validation/schemas");
const { serializeLabReport, serializeState } = require("../serializers");
const { newEventId, newReportId } = require("../utils/ids");

function mountLabReports(app) {
  const router = express.Router();

  router.get("/", (req, res) => {
    ok(res, {
      endpoints: {
        create: "POST /api/v1/lab-reports",
        list_for_batch: "GET /api/v1/lab-reports/batch/<batch_id>",
        get_one: "GET /api/v1/lab-reports/<report_id>",
      },
    });
  });

  router.post("/", requireRole("lab"), asyncHandler(async (req, res) => {
    const parsed = createLabReportSchema.safeParse(req.body ?? {});
    if (!parsed.success) {
      return error(res, "validation_error", "Invalid lab report data", 400, zodDetails(parsed.error));
    }
    const d = parsed.data;
    const batchId = d.batch_id;

    const [herb, state] = await Promise.all([
      prisma.herb.findUnique({ where: { batch_id: batchId } }),
      prisma.batchState.findUnique({ where: { batch_id: batchId } }),
    ]);
    if (!herb || !state) {
      return error(res, "not_found", `Batch '${batchId}' not found`, 404);
    }
    if (state.current_holder_id !== req.user.user_id) {
      return error(res, "forbidden", "Only the lab currently holding this batch can file a report", 403);
    }
    if (state.phase !== "at_lab") {
      return error(
        res,
        "invalid_state",
        `Lab reports can only be filed when phase is 'at_lab' (current: '${state.phase}')`,
        409
      );
    }

    const reportId = newReportId();
    const outcome = d.outcome;

    const { report, updatedState } = await prisma.$transaction(async (tx) => {
      const report = await tx.labReport.create({
        data: {
          report_id: reportId,
          batch_id: batchId,
          lab_id: req.user.user_id,
          test_type: d.test_type,
          test_date: d.test_date,
          results_summary: d.results_summary,
          outcome,
          certification_level: d.certification_level ?? null,
          purity_percentage: d.purity_percentage ?? null,
          moisture_content: d.moisture_content ?? null,
          ash_content: d.ash_content ?? null,
          heavy_metals_present: d.heavy_metals_present ?? null,
          pesticides_detected: d.pesticides_detected ?? null,
          active_compounds: d.active_compounds ?? null,
          potency_rating: d.potency_rating ?? null,
          report_url: d.report_url ?? null,
          notes: d.notes ?? null,
          recommendations: d.recommendations ?? null,
        },
      });

      const updatedState = await tx.batchState.update({
        where: { batch_id: batchId },
        data: { test_result: outcome, updated_at: new Date() },
      });

      await tx.batchEvent.create({
        data: {
          event_id: newEventId(),
          batch_id: batchId,
          event_type: "LAB_REPORT",
          actor_id: req.user.user_id,
          from_party_id: null,
          to_party_id: req.user.user_id,
          phase_before: state.phase,
          phase_after: state.phase,
          location: req.user.location,
          payload_json: {
            report_id: reportId,
            outcome,
            certification_level: d.certification_level ?? null,
          },
        },
      });

      return { report, updatedState };
    });

    return ok(res, { report: serializeLabReport(report), state: serializeState(updatedState) }, 201);
  }));

  router.get("/batch/:batch_id", requireAuth, asyncHandler(async (req, res) => {
    const herb = await prisma.herb.findUnique({ where: { batch_id: req.params.batch_id } });
    if (!herb) return error(res, "not_found", `Batch '${req.params.batch_id}' not found`, 404);

    const reports = await prisma.labReport.findMany({
      where: { batch_id: req.params.batch_id },
      orderBy: { created_at: "desc" },
    });
    return ok(res, { reports: reports.map(serializeLabReport), total: reports.length });
  }));

  router.get("/:report_id", requireAuth, asyncHandler(async (req, res) => {
    const report = await prisma.labReport.findUnique({ where: { report_id: req.params.report_id } });
    if (!report) return error(res, "not_found", `Lab report '${req.params.report_id}' not found`, 404);
    return ok(res, { report: serializeLabReport(report) });
  }));

  app.use("/api/v1/lab-reports", router);
}

module.exports = { mountLabReports };